library/std/src/net/addr.rs

```diff
@@ -291,9 +291,15 @@ impl SocketAddrV4 {
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn ip(&self) -> &Ipv4Addr {
         // SAFETY:
-        //   * `in_addr` is `struct { _: [u8; 4]; }` in most OSes.
         //   * `Ipv4Addr` is `struct { _: u32 }`.
-        //   * Both are `#[repr(C)]` and have the same size and alignment.
+        //   * `in_addr` is either:
+        //      * a `struct { _: u32; .. }` in POSIX-compliant OSes [1].
+        //      * a `struct { _: [u8; 4]; }` or `struct { _: [u16; 2]; }` in Win32 API [2][3].
+        //   * Both are `#[repr(C)]` and have the same size (4) and alignment (4)[3].
```

Comment:
`[u8; 4]` and `[u16; 2]` do not have the same alignment as `u32`. So if there is nothing else there the cast could be incorrect alignment-wise.

Reply:
In Windows, we take reference to the field from [sockaddr_in](https://docs.microsoft.com/en-us/windows/win32/api/winsock/ns-winsock-sockaddr_in):
```c
typedef struct sockaddr_in {
  short          sin_family;
  u_short        sin_port;
  struct in_addr sin_addr;
  char           sin_zero[8];
} SOCKADDR_IN, *PSOCKADDR_IN, *LPSOCKADDR_IN;
```

The reference to `sin_addr;` should be aligned to 4.

Reply:
What is making it 4-aligned, if `in_addr` is a u8/u16 array?

Certainly your comment in the code is factually wrong and also insufficient to ensure safety.

Reply:
Oops I was wrong. Is this a soundess bug in Windows?

Reply:
Indeed I think your PR introduces a soundness bug by increasing the alignment of `Ipv4Addr`.

Reply:
Oops. Sorry, was a wrong signal. The alignment of `in_addr` is at least 4.
Because `in_addr` contains an `unsigned long`:
```c
struct in_addr {
  union {
    struct {
      u_char s_b1;
      u_char s_b2;
      u_char s_b3;
      u_char s_b4;
    } S_un_b;
    struct {
      u_short s_w1;
      u_short s_w2;
    } S_un_w;
    u_long S_addr;
  } S_un;
};
```

Reply:
Still it might be safer to just define your "BE" type as `[u8; 4]`.

Reply:
Done.